### DISCUSSION TITLE
tasmota.get_power - BERRY

### DISCUSSION THREAD

**Discussion by mmmmm009** [arendst/Tasmota#20910] [Q&A] [answered]
Hi,

I'm trying to check if Power1 button on main page was pressed and set to "0".

```

class my_driver

def set_power_handler(cmd, idx)
print("Power changed")

   if (tasmota.get_power(0)) == false
      print("Power 1        changed")
   end

end

tasmota.add_driver(my_driver())                     
```

Whenever I press the button device resets.

What's wrong with this condition: ?
``` 
if tasmota.get_power(0) == false
    print("Power 1        changed")
end
```
THX!

**Comment by s-hadinger** [comment]
Without any details it's hard to guess. What is your version, configuration, logs...

**Comment by sfromis** [comment]
With a good config, the only issue with your code example is that it is bad syntax, missing `end` for the `class`.

A typical issue causing resets is bad config of a Button gpio input. If it is registering as "pushed" for 40 seconds, this triggers a recovery feature doing a full reset. You need to be sure if it is the normal (connecting to ground when pushing), or inverted (needs a `Button_i` config)

This is unrelated to the Tasmota web UI front page toggle, you cannot get anything directly from that, only indirectly by this toggling the power channel (typically a relay). The 40 seconds may "feel like" it is related to you using the web UI toggle, even if it is not.

Another issue may be if you have a relay drawing so much power that the ESP32 cannot run, and crashes due being starved of power.

**Comment by mmmmm009** [comment]
ESP32 with latest tamota32
Now I get the error:

```BRY: Exception> 'syntax_error' - input:12: unexpected token 'tasmota'```
(last line of the code)

**Reply by sfromis** [reply, marked as answer]
Yeah, that was the syntax error I just told you about :grin:

**Reply by mmmmm009** [reply]
THX!

**Reply by mmmmm009** [reply]
It's wired anyway, before I had 13.3.0 with TasmoCompiler and I didn't get syntax error msg but reset instead.
Now I flashed 13.4.0

**Reply by mmmmm009** [reply]
Is it possible that there is some kind of residual code in the Berry console and when I past a new-corrected code then I get wired behavior?

Is there any command resetting the console (cleaning all the scripts)?

**Reply by sfromis** [reply]
Whenever you execute that code piece again, you add a new driver instance, on top of keeping all the older ones active. You could:
`tasmota.remove_driver(global.myinstance) tasmota.add_driver(myinstance:=my_driver())`

If you use a build with `USE_BERRY_DEBUG`, you'll get a big red restart button in the Berry console page.

There's also the command `BrRestart`, but that does not work well if issued from Berry via `tasmota.cmd`.

**Comment by mmmmm009** [comment]
so... 

This part seems to reset tasmota `tasmota.set_power(3, false)`
When it's commented then it's ok.
Uncommented=reset

```
def set_power_handler(cmd, idx)
  print("Power changed")

  if tasmota.get_power(0) == false
	print("Power 1        changed")
	#tasmota.set_power(1, false)
	#tasmota.set_power(2, false)
	tasmota.set_power(3, false)
	gpio.set_pwm(27, 0)
	gpio.set_pwm(14, 0)
	gpio.set_pwm(12, 0)
  end

end
```

**Reply by sfromis** [reply]
Doing such functions from a callback can lead to race conditions, better defer the changes using `tasmota.set_power(0, ....)`

**Reply by mmmmm009** [reply]
You mean...
In `set_power_handler` check only `if tasmota.get_power(0) == false` and use some var here any true/false
and then in another funtion `tasmota.set_power(0, ....)` according to that variable ?

**Reply by sfromis** [reply]
Yeah, something like that, just avoid doing set calls from the callback.

Note that the `idx` parameter for `set_power_handler` also has bits about what power channels are on.

**Reply by mmmmm009** [reply]
Could idx[0] be a trigger for that new function?

**Reply by sfromis** [reply]
No. It is bits, not an array. `if idx&1` for first power channel (bit 0), 4 for third (bit 2), etc. Note that this is when it is on, not if it was changed. Same as with `tasmota.get_power`

**Reply by mmmmm009** [reply]
If this is my driver:
```
class my_driver

def set_power_handler(cmd, idx)
end

end
tasmota.add_driver(my_driver())                     
```

Then how can I check in another function if Power1 was set to 0 ?

I've tried:

```
def printIDX ()
  print(idx)
end
```

inside my_driver and outside. even if idx was a global var I wouldn't know how to use it to trigger a function.
Any help pls :)

**Reply by sfromis** [reply]
Well, if you want to pass idx to another function, have it as an argument in both the function call and definition. I'd prefer avoiding global vars as much as practical, they tend to become a mess when the code grows.
`tasmota.set_timer(0, /->printIDX(idx))`
or just
`printIDX(idx)`
...
`def printIDX(idx)`

And I already told you how you can mask the bit for checking the value.....

**Comment by mmmmm009** [comment]
So, If I have my_driver like this:

```
class my_driver

def set_power_handler(cmd, idx)
  print("Power changed")
end

def every_second()
end

def every_250ms()
end

def fast_loop()
end

end
tasmota.add_driver(my_driver())                     
tasmota.add_fast_loop(/-> my_driver.fast_loop()) 
```

whenever any Power is changed then idx is updated? (it is a global var?)

And now I need a function that is triggered by idx
the following function has idx as an argument but it is not triggered by idx change.
```
def printIDX(idx)
print(idx)
end
```
1. Should this function be inside my_driver?
2. I don't know how to build a function which will be triggered by idx change.

I've tried `tasmota.set_timer(0, /->printIDX(idx))` from Berrry concole but I get:
`BRY: Exception> 'syntax_error' - input:1: 'printIDX' undeclared (first use in this function)
`

**Reply by sfromis** [reply]
Yes, if you want to have a function as method in a class, put it inside the class. To call it, you'll have to refer to the class instance, like `self.printIDX(idx)`. This is not available from outside of the class, unless you have an instance reference kept somewhere.

Every time your `set_power_handler` callback is invoked, you have some change of power state, meaning that `idx` is different, and thus is already where you have a trigger point for `idx` being changed. Basically, just make the call from `set_power_handler`.

Note that what you get in the `idx` argument (not global) does not indicate which power channel was changed, only that some change happened. To "know" which channel, you can keep some current state, typically as an instance variable. Another approach for a single power channel could be:
`tasmota.add_rule('Power1#state', ...)`, where it is the change of that specific channel you get.

What I mentioned about using `tasmota.set_timer(0, ...)` to call something was specifically meant to avoid the problem that you should not to things like changing power state from code in or called from the `set_power_handler` callback, and no other commands there.

Instead of doing those `tasmota.add_` calls outside of the class, I'd recommend doing it from the `init` method which is called when a new instance is created, like:
```
tasmota.add_driver(self)                     
tasmota.add_fast_loop(/-> self.fast_loop()) 
``` 

In principle, you could also have `tasmota.add_fast_loop(self.fast_loop)`, but that would not allow the `fast_loop` method access to the instance, which makes things harder to build on.

Assuming that you may want to try iterative changes to the code, it can be useful to save current instance in a global var (one of the few cases where I'd recommend using global vars), to be able to use my previous suggestion of removing the old instance when adding a new.

**Reply by mmmmm009** [reply]
Thank you for your reply. 
Im still a beginner in all Tasmota stuff. I started from rules, then quick episode in Scripting and now trying Berry.

```
def powerOFF()
print("power off")
end

tasmota.remove_rule("Power1#state=0")
tasmota.add_rule("Power1#state=0", powerOFF)
```

This solution seems to be simple for me at this stage. But I'll be There soon ;)

**Reply by mmmmm009** [reply]
And something like this:
(based on:
_Originally posted by @LucidEye in https://github.com/arendst/Tasmota/discussions/17490#discussioncomment-6705322_
)

```
class MasterPower

def init()
	tasmota.add_driver(self)
end

def set_power_handler(cmd, idx)
	var pow = tasmota.get_power()
	if pow[1] == true
	    tasmota.set_timer(0, /-> tasmota.set_power(5, false))
	end
end

end
MasterPower()

```
could be another option? but it resets tasmota :/

**Reply by sfromis** [reply]
The problem with that technique is that `set_power_handler` will be called every time a power channel is set, even when it was not changed. This means that `tasmota.set_power(5, false)` will be executed every time, and will trigger `set_power_handler` to be called again, ad infinitum.

Using `set_power_handler` is not "an easy starting approach", but presumes that you're prepared to deal with all the interactions of multiple power channels, and take care of each situation without tripping on your own toes.

**Reply by mmmmm009** [reply]
ok. makes sense now
